core/src/ldml/ldml_processor.cpp

```diff
@@ -519,14 +524,40 @@ ldml_event_state::ldml_event_state(
   this->modifier_state = m;
   this->is_key_down    = i;
   this->event_flags    = e;
+
+  actions.persist_options = new km_core_option_item[1];
+  actions.persist_options[0] = NULL_OPTIONS[0];
+
+  // initialize actions
+  clear();
 }
 
 void ldml_event_state::commit() {
-  state->actions().commit();
+  // TODO-LDML: is this necessary? what's the lifecycle of actions.output?
+  actions.output = new km_core_usv[text.length()+1];
+  memcpy(actions.output, text.c_str(), text.length()+1);
+  // current implementation copies actions.output and doesn't retain it
+  state->set_actions(actions);
```

Comment:
`km::core::state::set_actions()` currently does NOT call `_actions.commit()` internally, but it kind of seems like it ought to. otherwise the action queue becomes unterminated.

Reply:
no, that's not right either, per the action_set_api.cpp test. It calls `set_actions()` and THEN pushes an `KM_CORE_IT_END` item (effectively `commit()`).. 

```c
void test_two_backspaces() {
  puts("test_two_backspaces");
  const km_core_action_item action_items[] = {
    bksp_action_item(KM_CORE_BT_CHAR, 0),
    bksp_action_item(KM_CORE_BT_CHAR, 0),
    end_action_item()
  };

  const km_core_actions actions = {
    2, // unsigned int code_points_to_delete;
    test_empty_output, // km_core_usv* output;
    test_env_opts, // km_core_option_item* persist_options;
    KM_CORE_FALSE, // km_core_bool do_alert;
    KM_CORE_FALSE, // km_core_bool emit_keystroke;
    KM_CORE_CAPS_UNCHANGED // new_caps_lock_state;
  };

  run_test(action_items, actions);
}
```

Reply:
My understanding is that the terminator is only applied on a retrieved list of actions, context items, etc. Internally, the terminator is not used when it is a std::list.

Reply:
@mcdurdin but.. core throws an assertion failure if the list isn't terminated.

Reply:
i'm calling actions().commit() at the end of the processing for now. PTAL if it needs to change

Reply:
Yep, for now that seems sensible. One day the need for it may disappear when the action queue goes away entirely.